### PR TITLE
[api] add error for chainlink service

### PR DESCRIPTION
**Pull request by Liuhaai**
Change the message of action error to be consistent geth' error (https://github.com/ethereum/go-ethereum/blob/master/core/error.go). 
This change enabled our chain compatible with external service , like Chainlink (https://github.com/smartcontractkit/chainlink/blob/7547b562579b4f201994fd85fb50d799ae1d918a/core/services/eth/errors.go#L72)

### CODE REVIEW

File: action/claimreward.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ func (c *ClaimFromRewardingFund) Cost() (*big.Int, error) {
 // SanityCheck validates the variables in the action
 func (c *ClaimFromRewardingFund) SanityCheck() error {
 	if c.Amount().Sign() < 0 {
-		return errors.Wrap(ErrBalance, "negative value")
+		return errors.Wrap(ErrNegativeValue, "negative value")
```
Comment (dustinxie) [inline action/claimreward.go:83]: directly return err, `ErrNegativeValue` same as "negative value"

File: action/depositreward.go [hunk 2]
```diff
@@ -80,7 +80,7 @@ func (d *DepositToRewardingFund) Cost() (*big.Int, error) {
 // SanityCheck validates the variables in the action
 func (d *DepositToRewardingFund) SanityCheck() error {
 	if d.Amount().Sign() < 0 {
-		return errors.Wrap(ErrBalance, "negative value")
+		return errors.Wrap(ErrNegativeValue, "negative value")
```
Comment (dustinxie) [inline action/depositreward.go:83]: directly return err, ErrNegativeValue same as "negative value"

File: action/protocol/generic_validator.go [hunk 3]
```diff
@@ -53,7 +53,7 @@ func (v *GenericValidator) Validate(ctx context.Context, selp action.SealedEnvel
 
 	pendingNonce := confirmedState.Nonce + 1
 	if selp.Nonce() > 0 && pendingNonce > selp.Nonce() {
-		return errors.Wrap(action.ErrNonce, "nonce is too low")
+		return errors.Wrap(action.ErrNonceTooLow, "nonce is too low")
```
Comment (dustinxie) [inline action/protocol/generic_validator.go:56]: directly return err, `ErrNonceTooLow` same as "nonce is too low"

File: action/transfer.go [hunk 4]
```diff
@@ -137,7 +137,7 @@ func (tsf *Transfer) Cost() (*big.Int, error) {
 func (tsf *Transfer) SanityCheck() error {
 	// Reject transfer of negative amount
 	if tsf.Amount().Sign() < 0 {
-		return errors.Wrap(ErrBalance, "negative value")
+		return errors.Wrap(ErrNegativeValue, "negative value")
```
Comment (dustinxie) [inline action/transfer.go:140]: same

File: action/transfer_test.go [hunk 5]
```diff
@@ -100,7 +100,7 @@ func TestTransfer(t *testing.T) {
 		tsf, err := NewTransfer(uint64(1), big.NewInt(-100), "2", nil,
 			uint64(100000), big.NewInt(0))
 		require.NoError(err)
-		require.Equal(ErrBalance, errors.Cause(tsf.SanityCheck()))
+		require.Equal(ErrNegativeValue, errors.Cause(tsf.SanityCheck()))
```
Comment (dustinxie) [inline action/transfer_test.go:103]: same

File: actpool/actpool.go [hunk 6]
```diff
@@ -202,7 +202,7 @@ func (ap *actPool) Add(ctx context.Context, act action.SealedEnvelope) error {
 	}
 	if ap.gasInPool+intrinsicGas > ap.cfg.MaxGasLimitPerPool {
 		actpoolMtc.WithLabelValues("overMaxGasLimitPerPool").Inc()
-		return errors.Wrap(action.ErrActPool, "insufficient gas space for action")
+		return errors.Wrap(action.ErrHitGasLimit, "insufficient gas space for action")
```
Comment (dustinxie) [inline actpool/actpool.go:205]: change `ErrHitGasLimit` to `errors.New("insufficient gas space for action")`? and directly return

File: actpool/actpool.go [hunk 7]
```diff
@@ -229,7 +229,7 @@ func (ap *actPool) Add(ctx context.Context, act action.SealedEnvelope) error {
 
 	caller := act.SrcPubkey().Address()
 	if caller == nil {
-		return errors.New("failed to get address")
+		return errors.Wrapf(action.ErrAddress, "failed to get address")
```
Comment (dustinxie) [inline actpool/actpool.go:232]: `return action.ErrAddress`

File: actpool/actpool.go [hunk 8]
```diff
@@ -366,8 +366,10 @@ func (ap *actPool) enqueueAction(addr address.Address, act action.SealedEnvelope
 		return errors.Wrapf(err, "failed to get sender's nonce for action %x", actHash)
 	}
 	confirmedNonce := confirmedState.Nonce
+	if actNonce <= confirmedNonce {
+		return errors.Wrapf(action.ErrNonceTooLow, "nonce too small, actNonce : %x", actNonce)
```
Comment (dustinxie) [inline actpool/actpool.go:370]: just `return action.ErrNonceTooLow`?

File: actpool/actqueue.go [hunk 9]
```diff
@@ -110,7 +110,7 @@ func (q *actQueue) Put(act action.SealedEnvelope) error {
 	if actInPool, exist := q.items[nonce]; exist {
 		// act of higher gas price cut in line
 		if act.GasPrice().Cmp(actInPool.GasPrice()) != 1 {
-			return errors.Wrap(action.ErrNonce, "gas price is smaller than the act of same nonce")
+			return errors.Wrap(action.ErrReplaceUnderpriced, "gas price is smaller than the act of same nonce")
```
Comment (dustinxie) [inline actpool/actqueue.go:113]: just `return action.ErrReplaceUnderpriced`

File: api/api.go [hunk 10]
```diff
@@ -416,18 +416,33 @@ func (api *Server) SendAction(ctx context.Context, in *iotexapi.SendActionReques
...
 		case action.ErrActPool:
 			desc = "Invalid actpool"
 		case action.ErrGasPrice:
 			desc = "Invalid gas price"
+		// Below errors are hardcoded for Chainlink Oracle
+		case action.ErrGasTooExpensive:
+			desc = `tx fee (0.1 ether) exceeds the configured cap (1.0 ether)`
+		case action.ErrExistedInPool:
+			desc = "known transaction"
+		case action.ErrReplaceUnderpriced:
```
Comment (dustinxie) [inline api/api.go:430]:
```
// these errors are specifically for chainlink error pattern match
case  action.ErrReplaceUnderpriced, action.ErrNonceTooLow ... :
    desc = err.Error()
```

File: api/api.go [hunk 11]
```diff
@@ -416,18 +416,33 @@ func (api *Server) SendAction(ctx context.Context, in *iotexapi.SendActionReques
 		}
 		var desc string
 		switch errors.Cause(err) {
-		case action.ErrBalance:
```
Comment (dustinxie) [inline api/api.go:419]: `action.ErrBalance` is no longer used? can delete if that's the case

File: actpool/actpool.go (hunk 8 above)
Comment (dustinxie) [reply actpool/actpool.go:370]: also please add a test case (if not yet) to cover this

File: api/coreservice.go [hunk 12]
```diff
@@ -289,18 +289,33 @@ func (core *coreService) SendAction(ctx context.Context, in *iotextypes.Action,
...
+		case action.ErrBalance:
+			desc = "insufficient balance for transfer"
+		case action.ErrNonce:
+			desc = "nonce too high"
+		case action.ErrAddress:
+			desc = "invalid sender"
+		case action.ErrHitGasLimit:
+			desc = "exceeds block gas limit"
+		case action.ErrNegativeValue:
+			desc = "negative value"
```
Comment (CoderZhi) [inline api/coreservice.go:318]: create a function `error description` to prevent duplicate code

**Review by CoderZhi — COMMENTED** [review]
same for others.

File: api/coreservice.go (hunk 12 above)
Comment (dustinxie) [reply api/coreservice.go:318]: +1

File: action/action.go [hunk 13]
```diff
@@ -118,9 +119,8 @@ func ClassifyActions(actions []SealedEnvelope) ([]*Transfer, []*Execution) {
 }
 
 func calculateIntrinsicGas(baseIntrinsicGas uint64, payloadGas uint64, payloadSize uint64) (uint64, error) {
-	if (math.MaxUint64-baseIntrinsicGas)/payloadGas < payloadSize {
-		return 0, ErrOutOfGas
+	if payloadGas == 0 || (math.MaxUint64-baseIntrinsicGas)/payloadGas < payloadSize {
```
Comment (dustinxie) [inline action/action.go:122]: `if math.MaxUint64-baseIntrinsicGas) < payloadGas * payloadSize`

File: api/api_test.go [hunk 14]
```diff
@@ -1273,15 +1273,15 @@ func TestServer_SendAction(t *testing.T) {
 	tests := []struct {
 		server func() (*Server, string, error)
 		action *iotextypes.Action
-		err    string
+		err    error
```
Comment (dustinxie) [inline api/api_test.go:1288]: can still use `err string`, see below

File: api/api_test.go [hunk 15]
```diff
@@ -1300,31 +1300,31 @@ func TestServer_SendAction(t *testing.T) {
 				return createServer(cfg, true)
 			},
 			testTransferPb,
-			"insufficient space for action: invalid actpool",
+			action.ErrTxPoolOverflow,
```
Comment (dustinxie) [inline api/api_test.go:1315]: `action.ErrTxPoolOverflow.Error()`

File: api/grpcserver_test.go [hunk 16]
```diff
@@ -1266,15 +1266,15 @@ func TestGrpcServer_SendAction(t *testing.T) {
 	tests := []struct {
 		server func() (*ServerV2, string, error)
 		action *iotextypes.Action
-		err    string
+		err    error
```
Comment (dustinxie) [inline api/grpcserver_test.go:1270]: can still use err string, see above comment

**Review by dustinxie — APPROVED** [review]

File: action/action.go [hunk 17]
```diff
@@ -83,22 +86,20 @@ func Verify(sealed SealedEnvelope) error {
...
 
 	h, err := sealed.envelopeHash()
 	if err != nil {
 		return errors.Wrap(err, "failed to generate envelope hash")
 	}
-	if sealed.SrcPubkey().Verify(h[:], sealed.Signature()) {
-		return nil
+	if !sealed.SrcPubkey().Verify(h[:], sealed.Signature()) {
+		log.L().Info("failed to verify action hash",
+			zap.String("act hash", hex.EncodeToString(h[:])),
```
Comment (dustinxie) [inline action/action.go:98]: "hash"

File: action/const.go [hunk 18]
```diff
@@ -9,30 +9,56 @@ package action
...
+	// ErrOversizedData is returned if the input data of a transaction is greater
+	// than some meaningful limit a user might use. This is not a consensus error
+	// making the transaction invalid, rather a DOS protection.
+	ErrOversizedData = errors.New("oversized data")
 )
+
+func LoadErrorDescription(err error) string {
+	switch errors.Cause(err) {
+	default:
+		return "Unknown"
```
Comment (dustinxie) [inline action/const.go:60]: move default: to end

**Review by raullenchai — APPROVED** [review]

**Pull request merged**